cmake/extensions.cmake

```diff
@@ -2261,3 +2263,368 @@ function(target_byproducts)
+    message(FATAL_ERROR "dt_num_regs(${ARGV0} ...) missing return parameter.")
+  endif()
+
+  foreach(arg ${req_single_args})
+    if(NOT DEFINED DT_REG_${arg})
+      message(FATAL_ERROR "dt_num_regs(${ARGV0} ...) "
+                          "missing required argument: ${arg}"
+      )
+    endif()
+  endforeach()
+
+  get_target_property(${var} devicetree_target "DT_REG|${DT_REG_NODE}|NUM")
+
+  set(${var} ${${var}} PARENT_SCOPE)
+endfunction()
+
+# Usage:
+#   dt_reg_addr(<var> NODE <node> [INDEX <idx>])
+#
+# Get the base address of the register block at index <idx>.
+# If INDEX is omitted, then the value at index 0 will be returned.
+#
+# the value will be returned in the <value> parameter.
+#
+# <var>          : Return variable where the address value will be stored
+# NODE <node>    : Node path
+# INDEX <idx>    : Index number
+function(dt_reg_addr var)
+  set(req_single_args "NODE")
+  set(single_args "INDEX")
+  cmake_parse_arguments(DT_REG "" "${req_single_args};${single_args}" "" ${ARGN})
+
+  if(${ARGV0} IN_LIST req_single_args)
+    message(FATAL_ERROR "dt_reg_addr(${ARGV0} ...) missing return parameter.")
+  endif()
+
+  foreach(arg ${req_single_args})
+    if(NOT DEFINED DT_REG_${arg})
+      message(FATAL_ERROR "dt_reg_addr(${ARGV0} ...) "
+                          "missing required argument: ${arg}"
+      )
+    endif()
+  endforeach()
+
+  if(NOT DEFINED DT_REG_INDEX)
+    set(DT_REG_INDEX 0)
+  endif()
+
+  get_target_property(${var}_list devicetree_target "DT_REG|${DT_REG_NODE}|ADDR")
+
+  list(GET ${var}_list ${DT_REG_INDEX} ${var})
+
+  if("${var}" STREQUAL NONE)
+    set(${var})
+  endif()
+
+  set(${var} ${${var}} PARENT_SCOPE)
+endfunction()
+
+# Usage:
+#   dt_reg_size(<var> NODE <node> [INDEX <idx>])
+#
+# Get the size of the register block at index <idx>.
+# If INDEX is omitted, then the value at index 0 will be returned.
+#
+# the value will be returned in the <value> parameter.
+#
+# <var>          : Return variable where the size value will be stored
+# NODE <node>    : Node path
+# INDEX <idx>    : Index number
+function(dt_reg_size var)
+  set(req_single_args "NODE")
+  set(single_args "INDEX")
+  cmake_parse_arguments(DT_REG "" "${req_single_args};${single_args}" "" ${ARGN})
+
+  if(${ARGV0} IN_LIST req_single_args)
+    message(FATAL_ERROR "dt_reg_addr(${ARGV0} ...) missing return parameter.")
+  endif()
+
+  foreach(arg ${req_single_args})
+    if(NOT DEFINED DT_REG_${arg})
+      message(FATAL_ERROR "dt_reg_addr(${ARGV0} ...) "
+                          "missing required argument: ${arg}"
+      )
+    endif()
+  endforeach()
+
+  if(NOT DEFINED DT_REG_INDEX)
+    set(DT_REG_INDEX 0)
+  endif()
+
+  get_target_property(${var}_list devicetree_target "DT_REG|${DT_REG_NODE}|SIZE")
+
+  list(GET ${var}_list ${DT_REG_INDEX} ${var})
+
+  if("${var}" STREQUAL NONE)
+    set(${var})
+  endif()
+
+  set(${var} ${${var}} PARENT_SCOPE)
+endfunction()
+
+# Usage:
+#   dt_has_chosen(<var> PROPERTY <prop>)
```

Comment:
Seems like we should have a `dt_has_prop` if needed and implement this using that.

Reply:
Do you want to ask whether a given node has a particular property ?
Like this:
```
dt_has_prop(result PATH <path> PROPERTY <prop>)
if(${result})
  dt_prop(value PATH <path> PROPERTY <prop>)
endif()
```

what's the value in the double call, the `dt_prop()` gives you everything in a single call.
```
dt_prop(value PATH <path> PROPERTY <prop>)
if(NOT DEFINED value)
  message("Node: ${path} does not have property ${prop}")
endif()

if(DEFINED value)
  message("Node: ${path} has property ${prop}, value is: ${value}")
endif()
```
with CMake, the return variable is `undefined` if the property doesn't exists for that node.
And if the property exists, the value of that property is returned.

So I really don't see the value in `dt_has_prop()`

Reply:
and of course if the property value is FALSE, test like:
```
dt_prop(value PATH <path> PROPERTY <prop>)
if(DEFINED value AND ${value})
  message("Node: ${path} has property ${prop}, and is TRUE")
endif()

if(DEFINED value AND NOT ${value})
  message("Node: ${path} has property ${prop}, and is FALSE")
endif()
```

Reply:
After thinking and discussing with @tejlmand, I'm resolving this "won't do" because this can't be implemented using dt_has_prop() with the current edtlib.

The reason why is that edtlib.Node.props only captures properties that are described in bindings. Since /chosen has no binding, it's not covered, and you have to use special-purpose edtlib APIs for /chosen. This is the case in C also; `DT_HAS_CHOSEN(prop)` in C is defined with a totally different set of generated macros.